### PR TITLE
Improve 'u' key in PCLVisualizer: search in shapes LUT too

### LINKED ISSUE #1240: Allow displaying LUT in PCLVisualizer when adding a vtkPolyData
Colors are nice (#1143) but the LUT cannot be displayed, when `U` is pressed in the visualizer:

```
Warning: In /build/buildd/vtk-5.8.0/Rendering/vtkScalarBarActor.cxx, line 270
vtkScalarBarActor (0x840a9a0): Need a mapper to render a scalar bar
```

A problem arise if the user has multiple `vtkPolyData` in the viewer and the LUTs are not the same
- Minimum/maximum scalar values are unlikely to be the same
- The user might have chosen 2 different LUTs for the `vtkPolyData`

https://github.com/PointCloudLibrary/pcl/blob/master/visualization/src/interactor_style.cpp#L1057

### PR DESCRIPTION
**Pull request by VictorLamoine**
Fixes #1240 

Features
===
1st commit
---
- [x] Remove `addModelFromPolyData(vtkPolyDataMapper)` because it is not needed (mapper can be set to a `vtkPolyData`) and causes some visualization functions to crash (setting shading for example)

2nd commit
---
- [x] The user can choose which cloud/shape LUT should be displayed with `setLookUpTableID`
- [x] LUT display is updated if the user triggers a `setShapeProperties`. The LUT is only updated if the shape updated is the one used for the LUT display (either because it is the one found automatically or the user defined which ID should be used)
- [x] The LUT is removed from the display if no cloud/shape was found with color information or if the user provided a wrong (= non existing or no color info) cloud/shape ID.

Test code
---
[#1241.zip](https://mega.nz/#!AAViHTDT!UTdNGHzptg3w143kTqjvtPkxTU81aEhDuHW1l7X_NwM)

Examples
---
capture du 2015-05-19 17 34 38
capture du 2015-05-19 17 34 56
capture du 2015-05-19 17 35 07
lut_display

### CODE REVIEW

**Comment by taketwo** [comment]
I wonder how you plan to implement the third item :)

**Comment by VictorLamoine** [comment]
I don't think that will be ever possible!

**Comment by VictorLamoine** [comment]
The code needs a lot of cleaning/refactoring but I think the idea is there.
I'm getting more and more keen on VTK, what a great library!

@taketwo [MP4 video (1.7 Mo)](https://mega.co.nz/#!EEsQ3SzQ!Qc4F4iraq0BeqzoW-Tz3Zc8w0iJVjl-qF2V7rKby6Ig
Inclure la clé du fichier)

**Comment by taketwo** [comment]
Smart!

**Comment by taketwo** [comment]
So you are now VTK and LUT expert, Victor... maybe you can address #1158 as well?

**Comment by VictorLamoine** [comment]
Because of #1244 I'm stuck.

**Comment by VictorLamoine** [comment]
Pull request/first message updated, please test with the example code provided!
I do not solve #1158 because it looks like a lot of work, but this pull request is clearly a step forward to solving it.

**Comment by taketwo** [comment]
I don't understand how your code makes sure that the LUT for the last cloud/shape added to the rendered is displayed. Can you explain the logic? Specifically, why you use forward iterators and `break` (not `continue`) inside the loops.

**Comment by VictorLamoine** [comment]
It's an error, it should be `continue`!

If no cloud/shape id is specified, we'll search for any cloud/shape that has a scalars to display the lookup table. If we found none, we should not display any lookup table. (`found_actor = false`)

When we exit the loop, if `actor_found` is set to true, we have found (at least) one actor with a lookup table so we will display this one, it will be either updated or added depending on `lut_enabled_` status.

I'm still updating this, it's not ready for merging but I would like this to be reviewed by other people.

**Comment by taketwo** [comment]
But since you want the last one added, shouldn't you rather iterate from the end to the beginning?

> I'm still updating this, it's not ready for merging but I would like this to be reviewed by other people.

To make the status of a pull request clear we may adopt a wide-spread convention: prefix the name with `[WIP]` if you are still working on the request, or `[RFC]` if you think you are done, but want comments.

**Comment by VictorLamoine** [comment]
> But since you want the last one added, shouldn't you rather iterate from the end to the beginning?

I'm not sure how I can achieve this, the `ShapeActor` typedef is a `boost::underdored_map`

```cpp
typedef boost::unordered_map<std::string, vtkSmartPointer<vtkProp> > ShapeActorMap;
```

The LUT displayed is the last found, not especially the last shape added.

**Comment by VictorLamoine** [comment]
@jspricke and @taketwo review would be appreciated :octopus: 

From what I've tested it works ok!

**Comment by taketwo** [comment]
I'm on vacation till the end of the month. Will have a look early next week.

File: visualization/include/pcl/visualization/interactor_style.h [hunk 1]
```diff
@@ -250,6 +260,11 @@ namespace pcl
           modifier_ = modifier;
         }
 
+        /** \brief ID used to fetch/display the look up table on the visualizer
+         * It should be set by PCLVisualizer \ref setLookUpTableID
+         * @note If empty, a random actor added to the interactor will be used */
+        std::string lut_actor_id_;
```
Comment (taketwo) [inline visualization/include/pcl/visualization/interactor_style.h]: Why public?

File: visualization/include/pcl/visualization/pcl_visualizer.h [hunk 2]
```diff
@@ -1872,6 +1860,12 @@ namespace pcl
         void
         setUseVbos (bool use_vbos);
 
+        /** \brief
+         * \param[in] id The id of the cloud/shape look up table to be displayed
+         * The look up table is displayed by pressing 'u' in the PCLVisualizer */
+        void
+		setLookUpTableID (const std::string id);
```
Comment (taketwo) [inline visualization/include/pcl/visualization/pcl_visualizer.h]: Tabs...

File: visualization/src/interactor_style.cpp [hunk 3]
```diff
@@ -1188,6 +1168,130 @@ pcl::visualization::PCLVisualizerInteractorStyle::OnKeyDown ()
...
+      if (!act->actor->GetMapper ()->GetLookupTable ())
+        continue;
+
+      if (!act->actor->GetMapper ()->GetInput ()->GetPointData ()->GetScalars ())
+        continue;
+
+      vtkScalarsToColors* lut = act->actor->GetMapper ()->GetLookupTable ();
+      lut_actor_->SetLookupTable (lut);
+      lut_actor_->Modified ();
+      actor_found = true;
```
Comment (taketwo) [inline visualization/src/interactor_style.cpp]: You should `break` here, otherwise search is continued even when a suitable actor is found.

File: visualization/src/interactor_style.cpp [hunk 4]
```diff
@@ -1188,6 +1168,130 @@ pcl::visualization::PCLVisualizerInteractorStyle::OnKeyDown ()
...
+    }
+
+    for (sm_it = shape_actors_->begin (); sm_it != shape_actors_->end (); ++sm_it)
```
Comment (taketwo) [inline visualization/src/interactor_style.cpp]: Don't start this look if `actor_found` is already `true`.

**Comment by taketwo** [comment]
> I'm not sure how I can achieve this, the `ShapeActor` typedef is a `boost::underdored_map`

You are right, sorry. Unordered map does not have a reverse iterator.

File: visualization/include/pcl/visualization/interactor_style.h (hunk 1 above)
Comment (VictorLamoine) [inline visualization/include/pcl/visualization/interactor_style.h]: I thought that a getter/setter would be add useless complexity so I made it public.
The PCLVisualizer directly modifies the string [here](https://github.com/PointCloudLibrary/pcl/pull/1241/files#diff-9c03ff6eb25cf5a57bad531b4ab8d9f3R4263) for example.

File: visualization/include/pcl/visualization/pcl_visualizer.h (hunk 2 above)
Comment (VictorLamoine) [inline visualization/include/pcl/visualization/pcl_visualizer.h]: Fixed

File: visualization/src/interactor_style.cpp (hunk 3 above)
Comment (VictorLamoine) [inline visualization/src/interactor_style.cpp]: The idea was to display the last actor found based on the principle that the last actor has many chances to be at the end of the `unordered_map` but after a short test it's not true.

So you are right, let's break the search and display the first actor with LUT support found.

**Comment by VictorLamoine** [comment]
Friendly ping :snail:

**Comment by VictorLamoine** [comment]
@taketwo Some free time for reviewing?

File: visualization/include/pcl/visualization/interactor_style.h (hunk 1 above)
Comment (taketwo) [inline visualization/include/pcl/visualization/interactor_style.h]: `PCLVisualizer` is a [friend](https://github.com/VictorLamoine/pcl/blob/visualizer_lut/visualization/include/pcl/visualization/interactor_style.h#L394), so will be able access this even if it is made private. I would strongly prefer to not have public member fields.

**Comment by taketwo** [comment]
For me this [last section](https://github.com/VictorLamoine/pcl/blob/visualizer_lut/visualization/src/interactor_style.cpp#L1276-L1293) is a bit hard to follow, e.g. it's not immediately obvious that all cases/possibilities are covered. But since I don't have any specific suggestions how it could be improved, and since it actually works (until proved otherwise :) ), I think I'm fine with merging this pull request. Please make a fix according to my inline comment though. Thanks!

**Comment by VictorLamoine** [comment]
`lut_actor_id_` is now private.

Yes this section is quite tricky and handles all the cases well from what I've tested.
I will write a tutorial about this later.
